_posts/2018-09-13-API-Gateway-Types-Compared.md

```diff
@@ -0,0 +1,153 @@
+
+The Private API Gateway is not publicly accessible or resolvable. Instead, it can only be accessed from within your VPC. To achieve this, AWS uses the [PrivateLink](https://aws.amazon.com/about-aws/whats-new/2017/11/introducing-aws-privatelink-for-aws-services/) technology. Diagram time!
+
+![Private API](/assets/posts/2018-09-13-API-Gateway-Types-Compared/private-api.png)
+
+AWS PrivateLink is an incredible addition to the AWS technology stack. It allows you to access many AWS Services (e.g. S3, KMS, SNS, CodeBuild and API Gateway) over a network interface in your own VPC. Take S3 as an example: before PrivateLink, any traffic from an EC2 instance to an S3 bucket would travel over the public internet. With PrivateLink, all traffic to S3 is routed to the Elastic Network Adapter(s) you've created in your own VPC. From this interface, the traffic will only traverse Amazon's internal network to S3. And as a bonus, the PrivateLink endpoint for S3 allows you to specify a policy which buckets can be accessed. This is a *great* security feature, because this effectively solves the problem of unauthorized data extraction to unknown buckets.
+
+Anyway, we digress. Back to API Gateway. By configuring an API Gateway as Private, it gets assigned a DNS name that can only be resolved from within the VPC. When you resolve the hostname for your API Gateway, the IP addresses for the PrivateLink Network Adapters are returned. Through this, all traffic to your API Gateway is routed over the AWS network.
+
+There is one caveat: if you try to access a Private API Gateway over VPN or VPC Peering, the hostname for the API Gateway will not resolve at the other end. There are two solutions for this:
+* Use the VPC Endpoint's DNS, which is publicly resolvable. However, you do need to specify the `Host` Header. An example in cURL: `curl https://vpce-04dee3730b094b205-zg5cu7h2.execute-api.eu-central-1.vpce.amazonaws.com/test -H 'Host: oxppfikjm4.execute-api.eu-central-1.amazonaws.com'`
+* Place an Application Load Balancer with an SSL certificate (e.g. `api.mydomain.com`) in front of the IP addresses of your PrivateLink network interfaces. Also deploy a custom domain name for `api.mydomain.com` and a base path mapping for your API Gateway. Then add a Route 53 record that points `api.mydomain.com` as an alias to your ALB. This solution is quite complex, but we've tested it and it works. We might write a separate blog post about it later.
```

Comment:
I would remove `We might write a separate blog post about it later.`

Reply:
Welll.. I think it's a bit weird to only explain half the solution, so I do want to say something. I changed it into

> Describing the full solution is outside the scope of this post, but we might write a separate blog post about it later.